### PR TITLE
set BaseFee to nil

### PR DESCRIPTION
**Pull request by NazariiDenha**
When computing the block hash, `l2geth` uses the whole block (including `BaseFee`). but the RPC client (e.g. SDK) does not receive `BaseFee` so the hash will be different.

we want to make sure we set `BaseFee` to `nil` inside `l2geth` too.

1. change it to `nil` in `worker.go` here: https://github.com/scroll-tech/go-ethereum/pull/206/files#diff-689874b30f6f905ce6c47cdefeddcf052b467a7936a981f4b6cf38939e10d924R933
2. do a `nil`-check everywhere `BaseFee` is used,

### CODE REVIEW

File: cmd/evm/internal/t8ntool/transition.go [hunk 1]
```diff
@@ -248,7 +248,7 @@ func Transition(ctx *cli.Context) error {
 	}
 	// Sanity check, to not `panic` in state_transition
 	if chainConfig.IsLondon(big.NewInt(int64(prestate.Env.Number))) {
-		if prestate.Env.BaseFee == nil {
+		if prestate.Env.BaseFee == nil && !(chainConfig.EnableEIP2718 && chainConfig.EnableEIP1559) {
```
Comment (colinlyguo) [inline cmd/evm/internal/t8ntool/transition.go:251]:
```suggestion
		if prestate.Env.BaseFee == nil && (chainConfig.EnableEIP2718 || chainConfig.EnableEIP1559) {
```

Comment (colinlyguo) [inline cmd/evm/internal/t8ntool/transition.go]:
```suggestion
		if prestate.Env.BaseFee == nil && chainConfig.EnableEIP2718 && chainConfig.EnableEIP1559 {
```

File: consensus/misc/eip1559.go [hunk 2]
```diff
@@ -72,6 +76,9 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header) *big.Int {
 		parentGasTargetBig       = new(big.Int).SetUint64(parentGasTarget)
 		baseFeeChangeDenominator = new(big.Int).SetUint64(params.BaseFeeChangeDenominator)
 	)
+	if !(config.EnableEIP2718 && config.EnableEIP1559) {
```
Comment (colinlyguo) [inline consensus/misc/eip1559.go]:
```suggestion
	if !config.EnableEIP2718 || !config.EnableEIP1559 {
```

File: core/state_transition.go [hunk 3]
```diff
@@ -250,7 +250,11 @@ func (st *StateTransition) preCheck() error {
 			}
 			// This will panic if baseFee is nil, but basefee presence is verified
 			// as part of header validation.
-			if st.gasFeeCap.Cmp(st.evm.Context.BaseFee) < 0 {
+			if st.evm.Context.BaseFee != nil && st.gasFeeCap.Cmp(st.evm.Context.BaseFee) < 0 {
+				return fmt.Errorf("%w: address %v, maxFeePerGas: %s baseFee: %s", ErrFeeCapTooLow,
+					st.msg.From().Hex(), st.gasFeeCap, st.evm.Context.BaseFee)
+			}
+			if st.evm.Context.BaseFee == nil && st.gasFeeCap.Cmp(big.NewInt(0)) < 0 {
```
Comment (colinlyguo) [inline core/state_transition.go:257]: we don't need this?

**Review by colinlyguo — DISMISSED** [review]
Some questions about the `if` condition, otherwise LGTM.

**Review by zimpha — DISMISSED** [review]

**Review by colinlyguo — DISMISSED** [review]

**Review by Thegaram — CHANGES_REQUESTED** [review]
Please update `genesis.go` so that we can initialize genesis with `BaseFee = nil`.

**Review by Thegaram — APPROVED** [review]

**Review by colinlyguo — APPROVED** [review]

**Review by icemelon — APPROVED** [review]

**Pull request merged**